src/Discord/Internal/Gateway/EventLoop.hs

```diff
@@ -182,11 +188,14 @@ runEventLoop thehandle sendablesData log = do loop
       Right (Hello _interval) -> do writeChan log "eventloop - unexpected hello"
                                     loop
       Right (HeartbeatRequest sq) -> do writeIORef (gatewayHandleLastSequenceId thehandle) sq
-                                        writeChan (librarySendables sendablesData) (Heartbeat sq)
+                                        sendHeartbeat sendablesData (gatewayHandleHeartbeatTimes thehandle) sq
                                         loop
       Right (InvalidSession retry) -> pure $ if retry then LoopReconnect else LoopStart
       Right Reconnect        -> pure LoopReconnect
-      Right HeartbeatAck     -> loop
+      Right HeartbeatAck     -> do
+        currTime <- getCurrentTime
+        result <- atomicModifyIORef' (gatewayHandleHeartbeatAckTimes thehandle) (dupe . const currTime)
+        result `seq` loop
```

Comment:
Doesn't `atomicModifyIORef'` already force the result, so we don't need the `seq`? I'm going off the docs for `atomicallyModifyIORef'` here, please correct me if wrong.

Reply:
it seq's whatever is put into the ioref with what it outputs, so it needs this output to be evaluated as well

Reply:
This is an incorrect understanding, and I can adjust the functions as a result; let me know if there's anything else to change!

Reply:
LGTM, but I can't do a local test for reasons, I trust you fully that it's fine though.